[MRG+1] Add average precision definitions and cross references

agitter opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->
Adds documentation requested in #4577 regarding pull request #9017.  Improved documentation may also address the concerns in #5379.

#### What does this implement/fix? Explain your changes.
The changes include:
- Remove both references to interpolated precision in the `average_precision_score` docstring
- Use a Wikipedia permalink for the AP reference in the `average_precision_score` docstring
- Copy the AP definition from `plot_precision_recall.py` to the `average_precision_score` docstring and `model_evaluation.rst`
- Recommend AP instead of trapezoidal AUC on PR curve operating points
- Add `average_precision_score` references to docstrings of related functions

#### Any other comments?
There has been some confusion about area under the curve for precision-recall curves in scikit-learn.  Currently, some documents refer to the AP computed by `average_precision_score` as the **area**.  Others refer to it as a **summarization**.  It would improve consistency to call AP the official area under the precision-recall curve in all documentation, but is that generally accepted?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

agitter committed:
Add average precision definitions and cross references

agitter commented:
Related to #9557 but there should not be any conflicts.

agitter committed:
Fix formatting error in AP formula

jnothman reviewed (commented):
Otherwise LGTM, thanks.

@amueller, this is also worth having in 0.19.1.

agitter commented:
I support #9586 for the sake of consistency.

@rhiever I'd appreciate your thoughts on phrasing here, especially:
> When summarizing a precision-recall curve, AP is preferable to computing the trapezoidal area under the operating points.

I also suggest that if we say AP is not equivalent to area under the curve, somewhere in the documentation we define what that area is and which scikit-learn function computes it.  AUCPR is a much more commonly reported metric than AP.

amueller commented on doc/modules/model_evaluation.rst:
```diff
@@ -636,7 +636,14 @@ by varying a decision threshold.
 (AP) from prediction scores. This score corresponds to the area under the
 precision-recall curve. The value is between 0 and 1 and higher is better.
-With random predictions, the AP is the fraction of positive samples.
+With random predictions, the AP is the fraction of positive samples. AP is
```
I would move the ``With random predictions`` sentence later. Otherwise it seem like it relates to the definition. I would mention here that there are interpolated variance and we're not doing that. Actually, I would rephrase the sentence above that this score corresponds to the area under the precision-recall curve. Because that is not entirely true / depends a lot on how you define that integral. Plotting the precision-recall curve, it's natural to do linear interpolation for visualization purposes (even though that might be misleading). But that's not how we calculate.

amueller replied:
Also add references to the "relationship of PR curve and ROC curve" paper and the IR book and the pascal VOC paper / code (saying it's different). And pointing out that linear interpolation leads to realizable classifiers for ROC but not PR curve.

amueller replied:
Maybe also add a reference to http://papers.nips.cc/paper/5867-precision-recall-gain-curves-pr-analysis-done-right.pdf because it also argues against linear interpolation.

agitter replied:
I support these changes and will make the revisions.

If we remove the AUCPR statement here, I recommend that somewhere we state what the canonical AUCPR calculation is in scikit-learn.  A user who wants to report AUCPR should be able to quickly find the appropriate function.  @amueller do you suggest that the `auc` function on the precision recall operating points is the best option?

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -69,7 +69,8 @@
 nth threshold. A pair :math:`(R_k, P_k)` is referred to as an
-*operating point*.
+*operating point*. When summarizing a precision-recall curve, AP is preferable
+to computing the trapezoidal area under the operating points.
```
Because linear interpolation does not correspond to a realizable classifier?

agitter replied:
My main gripe is the corner case where a classifier predicts only a single score, or scores that are close enough to be treated as equivalent.  Computing the area with the trapezoid rule in that case gives area > 0.5 to a poor classifier.  I've seen it cause problems for students trying to do automated hyperparameter searches based on AUCPR.

But whatever is going to be the reference AUCPR value should be recommended here.  If that's not AP, I'll change it.

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -149,13 +161,8 @@ def average_precision_score(y_true, y_score, average="macro",
-            <http://citeseerx.ist.psu.edu/viewdoc/
-            download?doi=10.1.1.157.5766&rep=rep1&type=pdf>`_
+           <http://en.wikipedia.org/w/index.php?title=Information_retrieval&
+           oldid=793358396#Average_precision>`_
```
why did you remove the references? I think these references should go into the user guide and we should explain exactly the relation between different approaches.

agitter replied:
I originally found it confusing to reference alternative implementations without explicitly stating they different from the AP implementation here.  I'll add them to the user guide.  I could also add them back here and present them as alternative interpolated approaches.

jnothman replied:
I agree. References in the implementation should be specific to the implementation. The user guide can give more commentary.

amueller reviewed (commented):
I think this would be good for 0.19.1. But I'm not entirely satisfied ;) It took me too long to figure out what's happening, I don't wish that on anyone else. Right not the docs still say we're computing the area under the PR curve, which is misleading. Also, we have less references now?

agitter committed:
Add precision recall references

agitter committed:
Describe precision recall references and interpolation

agitter committed:
Fix reference list formatting

agitter commented:
@amueller The latest commits add the references you suggested with some discussion, remove the area under the curve equivalence, and move the random performance comment.  You can [preview the changes here](https://13019-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/model_evaluation.html#precision-recall-and-f-measures).

I think the docs still need some work before merging.  The new version points out problems with linear interpolation but doesn't suggest a better approach.  I would like to add clear guidance about the recommended way to summarize a precision-recall curve.  `average_precision_score` and `auc` seem to be the options.

jnothman commented:
just leave auc out for now?

On 23 Aug 2017 4:24 am, "Anthony Gitter" <notifications@github.com> wrote:

> @amueller <https://github.com/amueller> The latest commits add the
> references you suggested with some discussion, remove the area under the
> curve equivalence, and move the random performance comment. You can preview
> the changes here
> <https://13019-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/model_evaluation.html#precision-recall-and-f-measures>
> .
>
> I think the docs still need some work before merging. The new version
> points out problems with linear interpolation but doesn't suggest a better
> approach. I would like to add clear guidance about the recommended way to
> summarize a precision-recall curve. average_precision_score and auc seem
> to be the options.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9583#issuecomment-324111573>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz65HzM630FXnjdZapIXKJhuddKVyNks5saxz6gaJpZM4O7vYk>
> .
>

agitter commented:
That would be fine to leave out AUC now and open a new issue to discuss AUCPR recommendations.  Should I remove the tentative recommendations to use AP from this pull request then?
- In `plot_precision_recall.py` I added "When summarizing a precision-recall curve, AP is preferable to computing the trapezoidal area under the operating points."
- In `auc` I added "For summarizing a precision-recall curve, see :func:`average_precision_score`."

jnothman commented:
What's wrong with recommending average precision? I hope we now believe
that what we've got is something worth using!

>

agitter commented:
I agree that the average precision is worth using.  I was getting confused about whether we should recommend AP as the preferred way to summarize a PR curve if many users will be accustomed to seeing AUCPR as the more common summarization.

If you think it reads well as-is, I'm happy with it too.

jnothman commented:
I think that depends entirely on which field you're coming from.

On 23 Aug 2017 9:40 pm, "Anthony Gitter" <notifications@github.com> wrote:

> I agree that the average precision is worth using. I was getting confused
> about whether we should recommend AP as the preferred way to summarize a PR
> curve if many users will be accustomed to seeing AUCPR as the more common
> summarization.
>
> If you think it reads well as-is, I'm happy with it too.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9583#issuecomment-324303147>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6yutPm2jqGWBRbF6nvUx9ZRKjadZks5sbA_BgaJpZM4O7vYk>
> .
>

agitter committed:
Change AUC to AP in precision recall example figure titles

agitter committed:
Present both AUC and AP as reasonable ways to summarize a PR curve

agitter commented:
> I think that depends entirely on which field you're coming from.

@jnothman Good point.  Commits f7e7cc1 and 169a2a6 are small changes to make it clear that both AP and AUC are reasonable ways to summarize a PR curve and AP != AUCPR.  I changed the title of the example precision-recall curves to call average precision AP instead of AUC.  The user guide can discuss the merits of the different options, including those not implemented in scikit-learn.

I have no more revisions planned.

jnothman commented:
thanks. I'll try to review shortly.

On 24 Aug 2017 1:11 am, "Anthony Gitter" <notifications@github.com> wrote:

> I think that depends entirely on which field you're coming from.
>
> @jnothman <https://github.com/jnothman> Good point. Commits f7e7cc1
> <https://github.com/scikit-learn/scikit-learn/commit/f7e7cc191a2795fcd66e1a6c3fce4a91383e3bd5>
> and 169a2a6
> <https://github.com/scikit-learn/scikit-learn/commit/169a2a68b39da89e6a3dc7b83954e79573d9be55>
> are small changes to make it clear that both AP and AUC are reasonable ways
> to summarize a PR curve and AP != AUCPR. I changed the title of the example
> precision-recall curves to call average precision AP instead of AUC. The
> user guide can discuss the merits of the different options, including those
> not implemented in scikit-learn.
>
> I have no more revisions planned.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9583#issuecomment-324367147>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz62dcHM4UtwQoZk2zY8_1QM9VdXsqks5sbEEEgaJpZM4O7vYk>
> .
>

agitter commented:
For review convenience, here are the previews of the docs I revised in the last two commits:
- [auc](https://13038-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.metrics.auc.html)
- [Precision-Recall examples](https://13038-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/auto_examples/model_selection/plot_precision_recall.html)

jnothman commented:
LGTM!

amueller commented on examples/model_selection/plot_precision_recall.py:
```diff
@@ -61,16 +61,20 @@
 *operating point*.
 
+AP and the trapezoidal area under the operating points
+(:func:`sklearn.metrics.auc`) are common ways to summarize a precision-recall
```
maybe emphasize that they lead to different results?

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -40,7 +40,9 @@ def auc(x, y, reorder=False):
     This is a general function, given points on a curve.  For computing the
-    area under the ROC-curve, see :func:`roc_auc_score`.
+    area under the ROC-curve, see :func:`roc_auc_score`.  For an alternative
+    way to summarize a precision-recall curve, see
```
I don't understand. AUC is area under the ROC curve, not PR curve.

agitter replied:
@amueller Per `This is a general function, given points on a curve` and the Davis and Goadrich paper, my understanding was that AUC can refer to trapezoidal area under any curve.  In the Davis and Goadrich terminology we can have AUC-ROC or AUC-PR.

What change do you recommend?

amueller replied:
never mind, I think your version is fine.

agitter replied:
I'm updating the `roc_auc_score`  description to say it computes ROC AUC (the acronym used in `_binary_roc_auc_score`) instead of the more general AUC.

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -108,6 +111,16 @@ def average_precision_score(y_true, y_score, average="macro",
+        \\text{AP} = \\sum_n (R_n - R_{n-1}) P_n
+
+    where :math:`P_n` and :math:`R_n` are the precision and recall at the nth
+    threshold [1]_.
```
Maybe explicitly say that this is not interpolated and that this is different from computing an interpolated area under the precision curve, which can be too optimistic.

amueller commented on sklearn/metrics/ranking.py:
```diff
@@ -396,6 +404,10 @@ def precision_recall_curve(y_true, probas_pred, pos_label=None,
 
+    See also
+    --------
+    average_precision_score : Compute average precision from prediction scores
```
maybe also add roc_curve? but also fine as-is.

amueller commented on doc/modules/model_evaluation.rst:
```diff
@@ -671,6 +684,24 @@ binary classification and multilabel indicator format.
+
+.. topic:: References:
+
+  * C.D. Manning, P. Raghavan, H. Schütze, `Introduction to Information Retrieval
```
Can you use the RST citation syntax and use references in the text, so that someone reading it knows what citation belongs to what statement? wiki and IR book should go to the definition. Then you can say "some areas such as computer vision use the interpolated area under the precision-recall curve instead [cite pascal VOC]. This can lead to an overly optimistic measure of classifier performance and has been criticized in the literature [Davis, Flach]. Right now, average_precision_score does not implement any interpolated variant".

agitter replied:
@amueller Yes, I can use RST citation syntax.  I haven't used it before, but [the docs](http://docutils.sourceforge.net/docs/user/rst/quickref.html#citations) make it look like I only need to add tags like `[Everingham2010]`.

I can also modify the text as suggested.  However, my understanding is that the Pascal VOC paper and Davis / Flach papers discuss two different concepts.  The VOC paper presents the 11-point interpolated average precision.  The Davis / Flach papers do not discuss that but instead criticize linear interpolation of a PR curve.

amueller replied:
VOC doesn't use 11 point in their code

agitter replied:
@amueller Commit 0c0f4bf046e501f63a24611f8a18d8fd220d62ad includes all of the changes you recommended during your last review.  I added citation links in `model_evaluation.rst` but used different phrasing.  My reading is that [Manning2008] and [Everingham2010] discuss interpolated AP whereas [Davis2006] and [Flach2015] criticize linear interpolation for AUC of precision-recall curves.

I'll confirm the citation links work once the docs finish building.

amueller reviewed (commented):
Some nitpicks. Looks pretty good overall!

agitter committed:
Add citation links to model evaluation guide and refine ranking function definitions

agitter committed:
Fix RST citation syntax

agitter commented:
This is ready for re-review.  The last commit fixed the RST citation links, which [you can preview here](https://13852-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/model_evaluation.html#precision-recall-and-f-measures).

jnothman reviewed (commented):
This is great, thanks!
